F401 false positive with future annotations

Not sure if this is already fixed in #992, but given the follow code, ruff incorrectly flags F401:

```python
from __future__ import annotations

from typing import Callable, Optional


def my_func(param: "Optional[Callable]" = None) -> "None":
    pass
```

```
Found 1 error(s).
example.py:5:20: F401 `typing.Callable` imported but unused
1 potentially fixable with the --fix option.
```

If `from __future__ import annotations` is not used, there are no errors.

This happens with Python 3.8, 3.9, 3.10, 3.11, and ruff 0.0.151

P.S. Thanks for making ruff so fast

--- a/resources/test/fixtures/pyflakes/future_annotations.py
+++ b/resources/test/fixtures/pyflakes/future_annotations.py
@@ -1,7 +1,7 @@
 from __future__ import annotations
 
 from dataclasses import dataclass
-from typing import List, Optional
+from typing import Callable, List, Tuple, Optional, Sequence
 
 from models import (
     Fruit,
@@ -37,4 +37,12 @@ def f(x: int) -> List[int]:
     return y
 
 
-x: Optional[int] = None
+x: Tuple[int, ...] = (1, 2)
+
+
+def f(param: "Optional[Callable]" = None) -> "None":
+    pass
+
+
+def f(param: Optional["Sequence"] = None) -> "None":
+    pass

--- a/src/check_ast.rs
+++ b/src/check_ast.rs
@@ -1191,7 +1191,10 @@ where
         let prev_in_literal = self.in_literal;
         let prev_in_type_definition = self.in_type_definition;
 
-        if self.in_type_definition && self.annotations_future_enabled {
+        if !(self.in_deferred_type_definition || self.in_deferred_string_type_definition)
+            && self.in_type_definition
+            && self.annotations_future_enabled
+        {
             if let ExprKind::Constant {
                 value: Constant::Str(value),
                 ..
@@ -2830,9 +2833,11 @@ impl<'a> Checker<'a> {
             self.scope_stack = scopes;
             self.parent_stack = parents;
             self.in_annotation = in_annotation;
+            self.in_type_definition = true;
             self.in_deferred_type_definition = true;
             self.visit_expr(expr);
             self.in_deferred_type_definition = false;
+            self.in_type_definition = false;
         }
     }
 
@@ -2861,9 +2866,11 @@ impl<'a> Checker<'a> {
             self.scope_stack = scopes;
             self.parent_stack = parents;
             self.in_annotation = in_annotation;
+            self.in_type_definition = true;
             self.in_deferred_string_type_definition = true;
             self.visit_expr(expr);
             self.in_deferred_string_type_definition = false;
+            self.in_type_definition = false;
         }
     }
 

